### PR TITLE
bpo-32494: Use gdbm_count for dbm_length if possible

### PR DESCRIPTION
**Pull request by corona10**
By this PR, we can use gdbm_count without exporting a new public API.

I ran the benchmark and it shows noticeable performance enhancement.
This can be measured by invalidating cached value.

## Benchmark 1
Run len(kv) after putting new value to invalidate the cache.

| Benchmark | [bpo-32494](https://bugs.python.org/issue32494)-master | [bpo-32494](https://bugs.python.org/issue32494)-proposed           |
+===========+==================+==============================+
| [bpo-32494](https://bugs.python.org/issue32494) | 262 us           | 42.2 us: 6.20x faster (-84%) |
+-----------+------------------+------------------------------+

```python
import pyperf

runner = pyperf.Runner()
runner.timeit(name="[bpo-32494](https://bugs.python.org/issue32494)",
              stmt="""
ret = len(kv)
kv[f'key-{ret}'] = f'value-{ret}'
"""
              ,
              setup = """
import dbm.gnu as gdbm
from test.support import TESTFN
kv = gdbm.open(TESTFN, 'c')
for i in range(1000):
    kv[f'key-{i}'] = f'value-{i}'
"""
              )


```
## Benchmark2

Remove caching code path to measure without putting new key/value.

```diff
-    if (dp->di_size < 0) {
+    if (1) {
```

+-----------+--------------------+-------------------------------+
| Benchmark | [bpo-32494](https://bugs.python.org/issue32494)-master-1 | [bpo-32494](https://bugs.python.org/issue32494)-proposed-1          |
+===========+====================+===============================+
| [bpo-32494](https://bugs.python.org/issue32494) | 109 us             | 590 ns: 185.32x faster (-99%) |
+-----------+--------------------+-------------------------------+

```python
import pyperf

runner = pyperf.Runner()
runner.timeit(name="[bpo-32494](https://bugs.python.org/issue32494)",
              stmt="""
ret = len(kv)
"""
              ,
              setup = """
import dbm.gnu as gdbm
from test.support import TESTFN
kv = gdbm.open(TESTFN, 'c')
for i in range(1000):
    kv[f'key-{i}'] = f'value-{i}'
"""
              )


```

https://bugs.python.org/issue32494

### CODE REVIEW

**Comment by corona10** [comment]
@pitrou Can you please take a look?

File: Modules/_gdbmmodule.c [hunk 1]
```diff
@@ -102,6 +102,20 @@ dbm_length(dbmobject *dp)
...
+        if (gdbm_count(dp->di_dbm, &count) == -1) {
+            if (errno != 0) {
+                PyErr_SetFromErrno(DbmError);
+            }
+            else {
+                PyErr_SetString(DbmError, gdbm_strerror(gdbm_errno));
+            }
+            return -1;
+        }
+        dp->di_size = count;
```
Comment (pitrou) [inline Modules/_gdbmmodule.c:121]: It seems this is blindly casting from `gdbm_count_t` to `int`. On my machine at least, `gdbm_count_t` is `unsigned long long int`. Even if overflow is rather theoretical here, I think we should check for it anyway.

**Review by pitrou — CHANGES_REQUESTED** [review]
Thank you for doing this. This looks mostly good, see below.

**Comment by corona10** [comment]
Thank you for the review :)

I have made the requested changes; please review again

**Review by pitrou — APPROVED** [review]
+1. Thanks for the `Py_ssize_t` fix!

**Pull request merged**